File: turbo_transformers/python/tests/bert_model_test.py
# Copyright (C) 2020 THL A29 Limited, a Tencent company.
# All rights reserved.
# Licensed under the BSD 3-Clause License (the "License"); you may
# not use this file except in compliance with the License. You may
# obtain a copy of the License at
# https://opensource.org/licenses/BSD-3-Clause
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" basis,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
# implied. See the License for the specific language governing
# permissions and limitations under the License.
# See the AUTHORS file for names of contributors.

import unittest
import torch
from transformers.models.bert.modeling_bert import BertModel, BertConfig
import numpy
import turbo_transformers
import sys
import os

sys.path.append(os.path.dirname(__file__))
import test_helper


class TestBertModel(unittest.TestCase):
    def init_data(self, use_cuda) -> None:
        torch.set_grad_enabled(False)
        torch.set_num_threads(4)
        turbo_transformers.set_num_threads(4)
        self.test_device = torch.device('cuda:0') if use_cuda else \
            torch.device('cpu:0')

        self.cfg = BertConfig()
        self.torch_model = BertModel(self.cfg)
        self.torch_model.eval()

        if torch.cuda.is_available():
            self.torch_model.to(self.test_device)

        self.turbo_model = turbo_transformers.BertModel.from_torch(
            self.torch_model, self.test_device, "turbo")

    def check_torch_and_turbo(self,
                              use_cuda,
                              batch_size,
                              seq_len,
                              use_memory_opt=True):
        self.init_data(use_cuda)
        num_iter = 1
        device_name = "GPU" if use_cuda else "CPU"
        input_ids = torch.randint(low=0,
                                  high=self.cfg.vocab_size - 1,
                                  size=(batch_size, seq_len),
                                  dtype=torch.long,
                                  device=self.test_device)

        torch_model = lambda: self.torch_model(input_ids)
        torch_result, torch_qps, torch_time = \
            test_helper.run_model(torch_model, use_cuda, num_iter)
        print(f'BertModel PyTorch({device_name}) QPS {torch_qps}')

        turbo_model = (lambda: self.turbo_model(input_ids))

        if use_memory_opt:
            turbo_transformers.bert_opt_mem_allocate_api(
                input_ids.size()[0],  # batch
                input_ids.size()[1],  # seq_len
                self.cfg.num_attention_heads,
                self.cfg.hidden_size,
                self.cfg.num_hidden_layers,
                "GPU" if 'cuda' in input_ids.device.type else "CPU")

        with turbo_transformers.pref_guard("bert_perf") as perf:
            turbo_result, turbo_qps, turbo_time = \
                test_helper.run_model(turbo_model, use_cuda, num_iter)
        print(f'BertModel TurboTransformer({device_name}) QPS {turbo_qps}')

        print(f"batch {batch_size} seq_len {seq_len}")
        print(torch.max(torch_result[0].cpu() - turbo_result[0].cpu()))
        self.assertTrue(
            numpy.allclose(torch_result[0].cpu(),
                           turbo_result[0].cpu(),
                           atol=1e-2,
                           rtol=1e-3))

    def bert_model_test_helper(self, use_memory_opt=False):
        if use_memory_opt:
            turbo_transformers.reset_allocator_schema("model-aware")

        for batch_size in [2, 4, 1]:
            for seq_len in [50, 4, 16]:
                if torch.cuda.is_available() and \
                        turbo_transformers.config.is_compiled_with_cuda():
                    self.check_torch_and_turbo(use_cuda=True,
                                               batch_size=batch_size,
                                               seq_len=seq_len,
                                               use_memory_opt=use_memory_opt)
                self.check_torch_and_turbo(use_cuda=False,
                                           batch_size=batch_size,
                                           seq_len=seq_len,
                                           use_memory_opt=use_memory_opt)

        if use_memory_opt:
            turbo_transformers.reset_allocator_schema("naive")

    def test_bert_model(self):
        # self.bert_model_test_helper(True)
        self.bert_model_test_helper(False)


if __name__ == '__main__':
    unittest.main()

File: example/cpp/bert_model_test.cpp
// Copyright (C) 2020 THL A29 Limited, a Tencent company.
// All rights reserved.
// Licensed under the BSD 3-Clause License (the "License"); you may
// not use this file except in compliance with the License. You may
// obtain a copy of the License at
// https://opensource.org/licenses/BSD-3-Clause
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" basis,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
// implied. See the License for the specific language governing
// permissions and limitations under the License.
// See the AUTHORS file for names of contributors.

#include "example/cpp/bert_model.h"
#include <cmath>
#include <future>
#include <thread>
#include <vector>

#include <iostream>
#include "catch2/catch.hpp"
#include "turbo_transformers/core/config.h"
#include "turbo_transformers/core/macros.h"

namespace turbo_transformers {
namespace loaders {
bool CheckCppBert(bool use_cuda, bool only_input) {
  BertModel model(model_file_path,
                  use_cuda ? DLDeviceType::kDLGPU : DLDeviceType::kDLCPU, 12,
                  12);
  std::vector<std::vector<int64_t>> position_ids{{1, 0, 0, 0}, {1, 1, 1, 0}};
  std::vector<std::vector<int64_t>> segment_ids{{1, 1, 1, 0}, {1, 0, 0, 0}};
  if (only_input) {
    position_ids.clear();
    segment_ids.clear();
  }
  auto vec = model({{12166, 10699, 16752, 4454}, {5342, 16471, 817, 16022}},
                   position_ids, segment_ids, PoolType::kFirst, false);
  REQUIRE(vec.size() == 768 * 2);
  // Write a better UT
  for (size_t i = 0; i < vec.size(); ++i) {
    REQUIRE(!std::isnan(vec.data()[i]));
    REQUIRE(!std::isinf(vec.data()[i]));
  }
  if (only_input) {
    std::cerr << vec.data()[0] << std::endl;
    REQUIRE(fabs(vec.data()[0] - -0.9791) < 1e-3);
    REQUIRE(fabs(vec.data()[1] - 0.8283) < 1e-3);
    REQUIRE(fabs(vec.data()[768] - 0.3837) < 1e-3);
    REQUIRE(fabs(vec.data()[768 + 1] - 0.1659) < 1e-3);
  } else {
    REQUIRE(fabs(vec.data()[0] - -0.3760) < 1e-3);
    REQUIRE(fabs(vec.data()[1] - 0.5674) < 1e-3);
    REQUIRE(fabs(vec.data()[768] - -0.2701) < 1e-3);
    REQUIRE(fabs(vec.data()[768 + 1] - 0.2676) < 1e-3);
  }
  return true;
}

bool CheckCppBertWithPooler(bool use_cuda, bool only_input) {
  BertModel model("models/bert.npz",
                  use_cuda ? DLDeviceType::kDLGPU : DLDeviceType::kDLCPU, 12,
                  12);
  std::vector<std::vector<int64_t>> position_ids{{1, 0, 0, 0}, {1, 1, 1, 0}};
  std::vector<std::vector<int64_t>> segment_ids{{1, 1, 1, 0}, {1, 0, 0, 0}};
  if (only_input) {
    position_ids.clear();
    segment_ids.clear();
  }
  auto vec = model({{12166, 10699, 16752, 4454}, {5342, 16471, 817, 16022}},
                   position_ids, segment_ids, PoolType::kFirst,
                   /*use_pooler*/ true);
  REQUIRE(vec.size() == 768 * 2);
  // Write a better UT
  for (size_t i = 0; i < vec.size(); ++i) {
    REQUIRE(!std::isnan(vec.data()[i]));
    REQUIRE(!std::isinf(vec.data()[i]));
  }
  if (only_input) {
    std::cerr << vec.data()[0] << std::endl;
    REQUIRE(fabs(vec.data()[0] - 0.9671) < 1e-3);
    REQUIRE(fabs(vec.data()[1] - 0.9860) < 1e-3);
    REQUIRE(fabs(vec.data()[768] - 0.9757) < 1e-3);
    REQUIRE(fabs(vec.data()[768 + 1] - 0.9794) < 1e-3);
  } else {
    REQUIRE(fabs(vec.data()[0] - 0.9151) < 1e-3);
    REQUIRE(fabs(vec.data()[1] - 0.5919) < 1e-3);
    REQUIRE(fabs(vec.data()[768] - 0.9802) < 1e-3);
    REQUIRE(fabs(vec.data()[768 + 1] - 0.9321) < 1e-3);
  }
  return true;
}

TEST_CASE("Bert", "Cpp interface") {
  CheckCppBert(false /*use_cuda*/, true /* only_input*/);
  CheckCppBert(false /*use_cuda*/, false /* only_input*/);
  if (core::IsCompiledWithCUDA()) {
    CheckCppBert(true /*use_cuda*/, true /* only_input*/);
    CheckCppBert(true /*use_cuda*/, false /* only_input*/);
  }
}

TEST_CASE("BertWithPooler", "Cpp interface") {
  CheckCppBertWithPooler(false /*use_cuda*/, false /* only_input*/);
  CheckCppBertWithPooler(false /*use_cuda*/, true /* only_input*/);
  if (core::IsCompiledWithCUDA()) {
    CheckCppBertWithPooler(true /*use_cuda*/, false /* only_input*/);
    CheckCppBertWithPooler(true /*use_cuda*/, true /* only_input*/);
  }
}

static std::vector<float> CallBackFunction(
    const std::shared_ptr<BertModel> model,
    const std::vector<std::vector<int64_t>> input_ids,
    const std::vector<std::vector<int64_t>> position_ids,
    const std::vector<std::vector<int64_t>> segment_ids, PoolType pooltype,
    bool use_pooler) {
  return model->operator()(input_ids, position_ids, segment_ids, pooltype,
                           use_pooler);
}

static bool test_multiple_threads(bool only_input, int n_threads) {
  std::shared_ptr<BertModel> model_ptr = std::make_shared<BertModel>(
      model_file_path, DLDeviceType::kDLCPU, 12, 12);
  std::vector<std::vector<int64_t>> input_ids{{12166, 10699, 16752, 4454},
                                              {5342, 16471, 817, 16022}};
  std::vector<std::vector<int64_t>> position_ids{{1, 0, 0, 0}, {1, 1, 1, 0}};
  std::vector<std::vector<int64_t>> segment_ids{{1, 1, 1, 0}, {1, 0, 0, 0}};
  if (only_input) {
    position_ids.clear();
    segment_ids.clear();
  }
  std::vector<std::thread> threads;
  threads.reserve(n_threads);

  std::vector<std::future<std::vector<float>>> result_list;
  result_list.reserve(n_threads);
  for (int i = 0; i < n_threads; ++i) {
    std::packaged_task<std::vector<float>(
        const std::shared_ptr<BertModel>,
        const std::vector<std::vector<int64_t>> &,
        const std::vector<std::vector<int64_t>> &,
        const std::vector<std::vector<int64_t>> &, PoolType, bool)>
        task(CallBackFunction);
    result_list.emplace_back(task.get_future());
    threads.emplace_back(std::thread(std::move(task), model_ptr, input_ids,
                                     position_ids, segment_ids,
                                     PoolType::kFirst, true));
  }

  for (int i = 0; i < n_threads; ++i) {
    auto vec = result_list[i].get();
    assert(vec.size() == 768 * 2);

    for (size_t i = 0; i < vec.size(); ++i) {
      assert(!std::isnan(vec.data()[i]));
      assert(!std::isinf(vec.data()[i]));
    }
    if (only_input) {
      assert(fabs(vec.data()[0] - 0.9671) < 1e-3);
      assert(fabs(vec.data()[1] - 0.9860) < 1e-3);
      assert(fabs(vec.data()[768] - 0.9757) < 1e-3);
      assert(fabs(vec.data()[768 + 1] - 0.9794) < 1e-3);
    } else {
      assert(fabs(vec.data()[0] - 0.9151) < 1e-3);
      assert(fabs(vec.data()[1] - 0.5919) < 1e-3);
      assert(fabs(vec.data()[768] - 0.9802) < 1e-3);
      assert(fabs(vec.data()[768 + 1] - 0.9321) < 1e-3);
    }
  }

  for (int i = 0; i < n_threads; ++i) {
    threads[i].join();
  }
  return true;
}
TEST_CASE("Bert-multiple-thread", "Cpp interface") {
  test_multiple_threads(false, 10);
  test_multiple_threads(true, 10);
}

}  // namespace loaders
}  // namespace turbo_transformers
